tools/linter_lib/custom_check.py

```diff
@@ -120,26 +120,13 @@
             "description": 'Avoid using "msgid" as a variable name; use "message_id" instead.',
         },
         {
-            "pattern": r"i18n\.t\([^)]+[^,\{\)]$",
-            "description": "i18n string should not be a multiline string",
-        },
-        {
-            "pattern": r"""i18n\.t\(['"].+?['"]\s*\+""",
-            "description": "Do not concatenate arguments within i18n.t()",
-        },
-        {
-            "pattern": r"""i18n\.t\([a-zA-Z]""",
-            "exclude": {"static/js/templates.js"},
-            "description": "Do not pass a variable into i18n.t; it will not be exported to Transifex for translation.",
-        },
-        {
-            "pattern": r"i18n\.t\(.+\).*\+",
+            "pattern": r"\$t\(.+\).*\+",
             "description": "Do not concatenate i18n strings",
         },
-        {"pattern": r"\+.*i18n\.t\(.+\)", "description": "Do not concatenate i18n strings"},
+        {"pattern": r"\+.*\$t\(.+\)", "description": "Do not concatenate i18n strings"},
         {
             "pattern": "[.]html[(]",
-            "exclude_pattern": r"""\.html\(("|'|render_|html|message\.content|util\.clean_user_content_links|i18n\.t|rendered_|$|\)|error_text|widget_elem|\$error|\$\("<p>"\))""",
+            "exclude_pattern": r"""\.html\(("|'|render_|html|message\.content|util\.clean_user_content_links|rendered_|$|\)|error_text|widget_elem|\$error|\$\("<p>"\))""",
```

Comment:
We should both audit the exclude list below as well as a similar rule for `.append`.

Reply:
I have *zero* confidence in anything this dumpster fire of a regex says or does not say. My strategy for this PR was to audit every interpolation by hand. I tried copying this regex for `append` in #17581 and it didn’t go well.

My future plan, for which I have a partial implementation I’m working on, is essentially to build a lightweight type checking system:

* Always store HTML in a variable or object property named `html` or `*_html`.
* When passing HTML to or from our own functions, always wrap it in an object expression with a property of such a name.
* Require the arguments for library function calls such as `jQuery#html` and `jQuery#append` (there are others) to be built only from other variables and properties of such names, Handlebars templates, and a small allow-list of safe escaping functions.
* Crucially, also require variables and properties of such names to be built only from other variables and properties of such names, Handlebars templates, and a small allow-list of safe escaping functions.
* We need to know the type of everything passed to the `$` constructor, so we make similar rules for everything else we need to pass to the `$` constructor: selector strings in `*_selector`, DOM elements in `*_element`, jQuery objects in `$*`. (Or we can peek at the TypeScript types for DOM elements and jQuery objects, once we have TypeScript.)
* Write an ESLint plugin to enforce all these rules at the AST level.
* Require any exceptions to be noted on a line-by-line basis with an `eslint-disable-line` comment, and don’t allow these to pass review without a very good reason.

Reply:
Yeah, this seems like a good long-term plan.